doc/design.rst

```diff
@@ -209,3 +209,237 @@ _________________
+Use poetry for project, build and depenency management.
+
+
+Code Formatting
++++++++++++++++
+
+**Description:**
+
+Currently we use Black and Isort for code formatting, though running them on a larger code base as pre-commit hooks or such can take quite a bit of time.
+
+**Downsides:**
+
+- Two tools and an aligned configuration of them are required to cleanly and correctly format the codebase.
+- Code needs to be processed at least twice as we apply two individual tools.
+- The performance of Black and Isort is okay but not great compared to other tools.
+
+**Rationale/History:**
+
+- Black and Isort have been used because they are battle-tested and widely used
+- When we opted for Black and Isort, ``ruff`` wasn't "a thing" yet and at best in its early stages.
+- Black and Isort already have been known by most python devs when we where selecting the tools
+
+**Ideas/Solutions:**
+
+As `Ruff <https://docs.astral.sh/ruff/>`_ is fairly stable and also tested and used by many Python projects
+we should consider transitioning to it.
+
+Advantages:
+
+- Well-tested
+- Widely used
+- Excellent performance
+- Single tool for imports and formatting the codebase
+- Simplifies adopting ruff for linting
+
+
+Pylint
+++++++
+
+**Description:**
+We are currently using Pylint instead of Ruff.
+
+**Downsides:**
+
+- Pylint is slower and less usable in pre-commit hooks
+- It is an additional tool, therefore at leas one more processing run of the code is required
+- No LSP (e.g. compare to `ruff lsp`)
+
+**Rationale/History:**
+
+- Well known
+- Pylint provides built-in project score/rating
+- Project score is good for improving legacy code bases which haven't been linted previously
+- Plugin support
+
+**Ideas/Possible Solutions:**
+
+Replacing Pylint with Ruff for linting would provide significant performance improvement. Additionally, Ruff offers an LSP and IDE integrations and is widely used these days. Additionaly there would be an additional synergy if we adopt ruff for formatting the code base.
+
+Transitioning to Ruff requires us to adjust the migration and improvement strategies for our projects:
+
+- Currently, our codebase improvements are guided by scores. However, with Ruff, a new approach is necessary. For example, we could incrementally introduce specific linting rules, fix the related issues, and then enforce these rules.
+
+- The project rating and scoring system will also need modification. One possiblity would be to run Ruff and Pylint in parallel, utilizing Pylint solely for rating and issue resolution while Ruff is incorporated for linting tasks.
+
+
+Security Linter
++++++++++++++++
+
+**Description:**
+As of today, the security linter does not fail if it has findings. This was intentionally done to simplify integration and adoption of the tool. Developers can still use the results to improve and find issues within the codebase, and additionally, a rating will be generated to provide some guidance on which projects need attention.
```

Comment:
This reminds me on the talk about defect dojo. https://media.ccc.de/v/38c3-vulnerability-management-with-defectdojo
I think, a future feature for the security linter could be that it creates a tickets if found on the main branch, but maybe this is actually not needed if we run it always and make it fail at some point.

Reply:
Good point but I think there is quite a bit of stuff to do before it generally can be enforced.

Reply:
agree